README.md
# niftygate - Drop-in Access Control via NFT Ownership

## The Pitch

Gate access to anything on the web, using any smart contract, deployed to any blockchain, without modifying your code.

## Overview

If you think of a token like a ticket to your club, NiftyGate would be the bouncer at the door.

- Compatible with the typical token standards, both fungible and not (ERC20, ERC721, ERC777, and ERC1155).
- Plays nice with any blockchain that speaks Web3 via WebSockets.
- Sits in front of anything that speaks HTTP.
- Has an embedded mode for use with client applications.
- Integrates with any auth system that supports HTTP headers.
- Comes with tools for choosing, deploying, debugging, and auditing compatible smart contracts.
- All shipped as single, self-contained binary.

- Cross-platform (tested on Linux and macOS, should work elsewhere).
- Open Source (MIT licensed), and written in Rust.

## What is it?

`niftygate` is a proxy for HTTP services, that validates signatures, providing
an AuthN layer for those services. It also has functionality for clients,
automatically injecting signatures into requests.

## What could I use this for?

As of `0.2`, Smart Contracts are supported. These allow using tokens as transferrable licenses to gated services.

Some examples:

- An OCI Registry that allowed licensed users to download an asset.
- An Issue Tracker that allowed project supporters to get priority support.
- A Project Roadmap that allowed customers proportional voting rights on feature development.
- A Deployment Pipeline that required at N-of-M authorized users to release to production.
- An App Store with decentralized licenses, where authors retain licensing control, not platforms.
- A Digital Library where lending is managed by tokens, and returns are enforced by Smart Contracts.

## Why does this exist?

This is really two questions, so let's break that down...

### Why is this concept useful?

Signature-based authentication avoids many of the pitfalls of password-based
authentication. The server only needs to know public keys, and if those are
leaked... who cares? They're _public_ keys.

Assuming a reasonable UI in both cases, users can rotate keys just as easily as
passwords. They can delegate a subset of permissions to other keys (just as we
do with access tokens elsewhere).

Keys are things users _generate_, not things users _memorize_. This removes
problems like password reuse entirely.

### Why is this approach practical?

Implementing this in the form of a proxy allows it to be applied to most existing
HTTP(S) applications without modification.

Separating the signing from the verification allows more fine-grained control
over what has access to private keys. It also means the proxy can be used only
where needed. If a client application can handle signing natively, then it should.

Implementing this as a set of composable middleware makes it simple to extend,
and (responsibly) experiment with new features.

It can be deployed as a sidecar container, if you're into that sort of thing.

## How do I obtain this majestic tool?

```shell
cargo install niftygate
```

Or, if you prefer Docker:

```shell
docker pull colstrom/niftygate:0.8.0
```

## How do I use it?

First, you'll need some sort of service you want to put the proxy in front of.

For demonstration and development purposes, `niftygate` includes an embedded
sample service that prints the request headers in the response body.

```shell
$ niftygate demo
tide::log Logger started
    level DEBUG
tide::server Server listening on http://127.0.0.1:8080
```

Next, you'll need to have an Ethereum JSON-RPC (Web3) service available for
`niftygate` to interact with. [Ganache](https://www.trufflesuite.com/ganache)
is an excellent option for local use, so the rest of this document assumes that.

First, pick the address you want to use for signing requests:
![screenshot showing how to choose an address in Ganache](assets/ganache-address.png)

Then, find the private key for this address:
![screenshot showing where to find a private key in Ganache](assets/ganache-private-key.png)

Add this private key to the environment:

Using `fish`:

```shell
set -x SECRET_KEY_DATA PUT_THE_DATA_HERE
```

Using vintage shells like `bash`:

```shell
export SECRET_KEY_DATA=PUT_THE_DATA_HERE
```

Now you're ready to run it! Let's have a look at the help:

```shell
$ niftygate web3 --help

niftygate-web3 0.1.0
Runs the proxy service

USAGE:
    niftygate web3 [FLAGS] [OPTIONS]

FLAGS:
    -h, --help                             Prints help information
    -V, --provides-account-verification    verify signatures and provide account addresses
    -B, --provides-balances                provide account balances
    -S, --provides-signatures              provide signatures
        --version                          Prints version information

OPTIONS:
        --address-header <name>        [env: ADDRESS_HEADER=]    [default: X-Web3-Account-Address]
    -b, --backend <url>                [env: BACKEND=]           [default: http://127.0.0.1:8080]
        --balance-header <name>        [env: BALANCE_HEADER=]    [default: X-Web3-Account-Balance]
        --balance-maximum <amount>     [env: BALANCE_MAXIMUM=]
        --balance-minimum <amount>     [env: BALANCE_MINIMUM=]
    -u, --balance-scale <unit>         [env: BALANCE_SCALE=]     [default: Wei]
    -c, --challenge <phrase>           [env: CHALLENGE=]         [default: totes-legit]
    -l, --listen <address>             [env: LISTEN=]            [default: 0.0.0.0:8000]
    -K, --secret-key-data <hex>        [env: SECRET_KEY_DATA=]
    -k, --secret-key-file <path>       [env: SECRET_KEY_FILE=]
        --signature-header <name>      [env: SIGNATURE_HEADER=]  [default: X-Web3-Signature]
    -w, --web3-rpc-url <url>           [env: WEB3_RPC_URL=]      [default: ws://127.0.0.1:7545]
```

All of those options are optional, though it doesn't **do** anything unless you
enable at least one of them. Let's focus on two scenarios.

### Scenario 1 - Providing Signatures

In this mode, `niftygate` adds a signature to every request it handles.

```shell
$ niftygate web3 --provides-signatures
tide::log Logger started
    level DEBUG
tracing::span new
tracing::span build
isahc::agent agent waker listening on 127.0.0.1:61558
isahc::agent agent_thread; port=61558
isahc::agent agent took 409.705µs to start up
tide::server Server listening on http://0.0.0.0:8000
```

A secret (private) key is needed to sign these requests, though. It can be given
as a file (`--secret-key-file`), or a hex string (`--secret-key-data`). If both
are given, `--secret-key-file` is ignored. These can also be given via
environment, which is exactly what we did above.

This key is used to sign a challenge phrase (currently given by `--challenge`),
which should be known to both client and server. It does not need to be private.

This signature is added to the request using the header given by `--signature-header`.

### Scenario 2 - Providing Account Verification

In this mode, `niftygate` expects requests to contain a signature header, and
will reject requests that do not contain one. These headers will be verified,
and the address used to sign them will be added to each request it handles.

Note that the client does not explicitly provide an address. They sign a
challenge phrase that is known to both signer and verifier, and the signing
address is recoverable from a valid signature.

```shell
$ niftygate web3 --provides-account-verification
tide::log Logger started
    level DEBUG
tracing::span new
tracing::span build
isahc::agent agent waker listening on 127.0.0.1:56181
isahc::agent agent_thread; port=56181
isahc::agent agent took 312.747µs to start up
tide::server Server listening on http://0.0.0.0:8000
```

If the signature is valid, the address is added to the request using the header
given by `--address-header`. It is assumed that the backend knows what to do
with this header. For instance, you could use the `X-Remote-User` header to
integrate `niftygate` with [Dex](https://dexidp.io/docs/connectors/authproxy/).

### Scenario 3 - Providing Account Balances

In this mode, `niftygate` expects requests to contain an account header, and
will reject requests that to not contain one. The balance of the address given
in this header will be added to each request it handles.

```shell
$ niftygate web3 --provides-balances
tide::log Logger started
    level DEBUG
tracing::span new
tracing::span build
isahc::agent agent waker listening on 127.0.0.1:54608
isahc::agent agent_thread; port=54608
isahc::agent agent took 313.524µs to start up
tide::server Server listening on http://0.0.0.0:8000
```

By default, this doesn't care _what_ the balance is, it just adds it to the
header given by `--balance-header`. However, there are a few options to restrict
access based on the balance.

The `--balance-minimum` option can be used to require a minimum balance. This
can be used as a form of "skin in the game", by requiring a non-zero balance
(which may be useful in to prevent bots or assist rate limiting), or to gate
access to only high rollers if you wanted to do that for some reason.

The `--balance-maximum` option can be used to reject accounts with a balance
over a given amount. This can be used to limit risk, or gate access to only the
common folk if you were so inclined.

The `--balance-unit` option can be used to conveniently scale the above limits.
How many zeroes are there in a Gwei? No need to remember, just set
`--balance-unit=Gwei` and `niftygate` will do the scaling internally. The
balance header will still be given in Wei, though.

## Notes

The examples above show demonstrate using the features independently, but they
are all implemented as composable middleware. You can enable all of them in a
single process. It seems a bit silly to use both signing and verifying in the
same process, but there's no reason you **can't** do it.

## Try it out!

Assuming you've enabled all the features, and the proxy is in front of the demo
application, if you were to query each with `curl`, you'd see something like the
following:

Querying the demo app directly:

```shell
$ curl -s http://127.0.0.1:8080 | sort
accept: ["*/*"]
host: ["127.0.0.1:8080"]
user-agent: ["curl/7.64.1"]
```

Querying through the proxy:

```shell
$ curl -s http://127.0.0.1:8000 | sort
accept-encoding: ["deflate, gzip"]
accept: ["*/*"]
content-length: ["0"]
content-type: ["application/octet-stream"]
forwarded: ["by=127.0.0.1:8000;for=\"127.0.0.1:52698\";host=127.0.0.1:8000;proto=http"]
host: ["127.0.0.1:8000"]
user-agent: ["curl/7.64.1"]
x-web3-address: ["24f9f97c9e540fed57ef81f6c9aeabdb6fc73acd"]
x-web3-balance: ["100000000000000000000"]
x-web3-signature: ["krpQZO9WpgAEso2uk6eAKDy29QjeVtr+gdDZ7iG4bFkYiTfNvTvU5l4bb2iod5F4Ab8a8tJqzXHSXLkyz9U/gRs="]
```

## Smart Contracts

### Embedded Presets

To help you get started, `niftygate` embeds a set presets from the OpenZeppelin
project. These address some of the more common use cases, and are (according to
the OpenZeppelin documentation) production-ready.

Currently, there are five presets included, along with utilities for deploying
and interacting with each. These can be found under the `contract` subcommand.

To help you figure out which contract is appropriate, check the `guide`
subcommand. There you'll find an interactive program that will ask you a few
simple questions, and present you with a recommendation.

## Roadmap

- 0.1 - Initial Release with Web3 Account Verification
- 0.2 - Support Token Standards
  - Support for [ERC-20 (Token Standard)](https://eips.ethereum.org/EIPS/eip-20)
  - Support for [ERC-777 (Token Standard)](https://eips.ethereum.org/EIPS/eip-777)
  - Support for [ERC-721 (Non-Fungible Token Standard)](https://eips.ethereum.org/EIPS/eip-721)
  - Support for [ERC-1155 (Multi Token Standard)](https://eips.ethereum.org/EIPS/eip-1155)
- 0.3 - Observability for Smart Contracts
- 0.5 - Support Upgradeable Smart Contracts
- 0.6 - Renamed from sig-proxy to NiftyGate
- 0.7 - TLS Support
- 0.8 - Asset Management and (experimental) marketplace support

## Wishlist (in no particular order)

Infura Integration

Support for [RFC-7486 (HTTP Origin-Bound Authentication)](https://tools.ietf.org/html/rfc7486)

- Replacing fixed messages with random challenges, adding nonces and expiration, etc.

Optional endpoint to serve an embedded script for in-browser support.

- Use [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) for key generation and signing.
- Use [Web Storage API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API) for key storage.
- Use [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) for header manipulation and auth flow.

X509 Support

- This would provide an alternative where a chain-of-trust model is preferred.

PGP Support

- This would provide an alternative where a web-of-trust model is preferred.

## License

`niftygate` is available under the MIT License. See `LICENSE.txt` for the full text.

`niftygate` embeds contract ABI specs, and documentation excerpts from [OpenZeppelin Contracts](https://github.com/OpenZeppelin/openzeppelin-contracts).
These are also distributed under the terms of the MIT License [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.1.0/LICENSE).

## How can NiftyGate help me?

### I already have a Site, and a Smart Contract

Run niftygate in server mode, give it the address of your site, and the address of your contract.

### I have a site, but I don't want to use Smart Contracts

Run niftygate in server mode, Give niftygate the address to your site, and it will let users log in with their Wallet.

### I have a non-web application, and I want to interface with Smart Contracts

Compile niftygate as a library, and embed it in your application.

### I have a non-web application, and I don't want to modify my code.

Run niftygate in client mode, and point your application at it.

### I'm using a container platform, like Docker or Kubernetes.

Run niftygate in either client or server mode, deployed as a sidecar container.

### I'm using Kubernetes, and I want to integrate Smart Contracts into a Controller.

Use the `contract events --stream` mode. This will output matching events as compact JSON on STDOUT.

### I don't have a Smart Contract, but I want one.

Deploy one of the built-in contracts that comes with niftygate, no code required!

### I don't know what kind of Smart Contract I need.

Use the interactive guide built into niftygate.

### I don't have an Ethereum node to connect to.

Sign up for an account at infura.io, and point niftygate at it. Free accounts
allow up to 100k daily requests, and if you need more than that, they have paid
plans available.

### I want to use a private blockchain.

Go for it. Most of the development of niftygate is done on a private blockchain.
As long as it speaks JSON-RPC + Web3 over WebSockets, it should be fine.

Ganache and geth are good options for this, depending on your needs.

### I want to use my own identity platform.

Run niftygate in server mode, and tell it which headers it should pass along.


CHANGELOG.md
# 0.x (unreleased)

- (internal) update rust edition to 2021
- (dependency) camino 1.0.7 -> 1.0.8
- (dockerfile) added aarch64 support
- (dockerfile) added static linking

# 0.8.0

- (feature) crowdfunding support (experimental)
- (feature) marketplace support (experimental)
- (feature + commandline) added "asset" subcommand
  - (package) this is also available standalone via the "niftygate-asset" crate.
  - provides discovery of Solidity compiler versions.
  - can download any version of Solidity with no external tooling required.
    - provides transparent on-disk compression for cached assets.
      - supports LZ4, Deflate (gzip), and Brotli, depending on your needs.
    - integrity verification, to facilitate secure supply chains.
  - flexible caching policies to support environments like CI/CD, air-gapped, or immutable deployments.
  - granular network access policies to reduce attack surface when downloading.
  - supports custom mirrors, for trusted build environments.
  - respects various platform-specific standards for paths
  - automation and scripting-friendly.
    - provides built-in tooling for cache maintenance (usage reporting, cleanup, etc).
    - built-in filtering with support for Regular Expressions and Semantic Versioning constraints.
  - user-friendly interative prompts to guide you (optional, of course).
  - (experimental) able to extract embedded WebAssembly from soljson release artifacts.
  - (unsupported) built-in benchmarking of supported compression schemes.
  - (unsupported) built-in tooling for analyzing soljson release artifacts.
- (fix) CLI enforces required arguments for TLS support.
- (internal) replace a fallible conversion with an infallible one.
- (feature) added TLS support to embedded demo application.
- (package) "guide" subcommand moved into "niftygate-guide" crate.
- (package) "contract" subcommand moved into "niftygate-contract" crate.
- (niftygate-certificate) now with more accurate help text!
- (commandline) "units" subcommand can now convert values.
  - this introduces a "units convert" subcommand for use in scripting.
  - the previous functionality is still available via "units show".
- (library) BalanceScale now implements `Copy`.
- (deprecation) WrappedError is deprecated, use anyhow::Error.
- (deprecation) WrappedResult is deprecated, use anyhow::Result.
- (deprecation) command::run() is deprecated, use command::Command.execute().
- (deprecation) command::HexData is deprecated as an import path. It has been moved to the crate root.
- (breaking) rename command::Options to command::Command.
- (breaking) command::Options::Web3 enum variant moved to struct command::web3::Command.
- (breaking) command::Options::Demo enum variant moved to struct command::demo::Command.
- (breaking) command::Options::Units enum variant moved to struct command::units::Command.
- (breaking: niftygate-certificate) command::Generate takes time::OffsetDateTime now, now chrono::DateTime<Utc>
  - this is due to a change in a dependency, not an active choice to move away from Chrono.
- (dependency) anyhow 1.0.43 -> 1.0.57
- (dependency) async-std 1.10.0 -> 1.11.0
- (dependency) async-trait 0.1.51 -> 0.1.53
- (dependency) blake3 1.2.0 -> 1.3.1
- (dependency) brotli 3.3.2 -> 3.3.4
- (dependency) console 0.14.1 -> 0.15.0
- (dependency) dialoguer 0.9.0 -> 0.10.1
- (dependency) digest 0.10.1 -> 0.10.3
- (dependency) ethcontract 0.15.3 -> 0.17.0
- (dependency) ethcontract-generate 0.15.3 -> 0.17.0
- (dependency) heck 0.3.3 -> 0.4.0
- (dependency) pem 0.8.3 -> 1.0.2
- (dependency) rcgen 0.8.13 -> 0.9.2
- (dependency) regex 1.5.4 -> 1.5.5
- (dependency) secp256k1 0.20.3 -> 0.21.3
- (dependency) semver 1.0.4 -> 1.0.9
- (dependency) serde 1.0.130 -> 1.0.137
- (dependency) serde_json 1.0.67 -> 1.0.81
- (dependency) structopt 0.3.23 -> 0.3.26
- (dependency) strum 0.21.1 -> 0.24.0
- (dependency) surf 2.3.1 -> 2.3.2
- (dependency) thiserror 1.0.29 -> 1.0.31
- (dependency) wasm-bindgen-futures 0.4.28 ->? 0.4.29
- (dependency) web3 0.17.0 -> 1.18.0
- (dependency) x509-parser 0.11.0 -> 0.13.1
- (dependency) xdg 2.4.0 -> 2.4.1

# 0.7.0

- (feature) add TLS support!
- (commandline) added --with-tls option.
- (commandline) added --tls-certificate-path option.
- (commandline) added --tls-key-path option.
- (feature + commandline) added "certificate" subcommand.
  - (package) this is also available standalone via the "niftygate-certificate" crate.
  - can generate self-signed certificates, CA certs, and can sign certs.
  - this provides a way to generate and sign certificates without external tools.
  - includes support for constrained authorities
    - can restrict the number of intermediate authorities allowed.
    - can limit the scope of the authority based on DNS, IP, Directory Name, or Email.
    - constraints can be combined, and exclusions are supported.
- (lint) fixed collapsible_if(s)
- (lint) fixed needless_borrow(s)
- (lint) fixed redundant_static_lifetimes(s)
- (depencency) added anyhow
- (dependency) added thiserror
- (dependency) added tide-rustls

# 0.6.3

- (dependency) async-std 1.9.0 -> 1.10.0
- (dependency) ethcontract 0.12.2 -> 0.15.3
- (dependency) ethcontract-generate 0.12.2 -> 0.15.3
- (dependency) serde 1.0.129 -> 1.0.130
- (dependency) serde_json 1.0.66 -> 1.0.67
- (dependency) structopt 0.3.22 -> 0.3.23
- (dependency) surf 2.2.0 -> 2.3.1
- (dependency) web3 0.16.0 -> 0.17.0
- (internal) regenerate bindings

# 0.6.2

- (internal) split into packages.
  - niftygate is the main package, containing the app and library.
  - niftygate-bindings contains the bindings for various smart contracts.
  - niftygate-bindgen contains the code generator for the contract bindings.
  - there should be no impact to compatibility with this change.
  - this resolves an ongoing challenge with dependency upgrades.

# 0.6.1


src/command.rs
use crate::{middleware::ethereum::*, WrappedResult};
use secp256k1::SecretKey;
use std::{fs, path::PathBuf, str::FromStr};
use structopt::StructOpt;
use strum::VariantNames;
use tide::{
  http::{headers::HeaderName, Url},
  log,
};
use web3::types::{Address, U256};

mod contract;
mod guide;

#[derive(Debug)]
pub struct HexData(Vec<u8>);

impl FromStr for HexData {
  type Err = hex::FromHexError;
  fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
    hex::decode(s).map(Self)
  }
}

#[derive(Debug, StructOpt)]
#[structopt()]
pub enum Options {
  #[structopt(about = "Runs a sample app that returns request headers in the response body")]
  Demo {
    #[structopt(
      env,
      long,
      short,
      value_name = "address",
      default_value = "127.0.0.1:8080"
    )]
    listen: String,
  },
  #[structopt(about = "Prints a table of recognized units and scaling values")]
  Units {},
  #[structopt(about = "Runs the proxy service")]
  Web3 {
    #[structopt(
      env,
      long,
      short,
      value_name = "address",
      default_value = "0.0.0.0:8000"
    )]
    listen: String,

    #[structopt(
      env,
      long,
      short,
      value_name = "url",
      default_value = "http://127.0.0.1:8080"
    )]
    backend: Url,

    #[structopt(
      env,
      short,
      long,
      value_name = "url",
      default_value = "ws://127.0.0.1:7545"
    )]
    web3_rpc_url: Url,

    #[structopt(
      env,
      long,
      value_name = "name",
      default_value = "X-Web3-Account-Address"
    )]
    address_header: HeaderName,

    #[structopt(
      env,
      long,
      value_name = "name",
      default_value = "X-Web3-Account-Balance"
    )]
    balance_header: HeaderName,

    #[structopt(
      env,
      long,
      value_name = "name",
      default_value = "X-Web3-ERC1155-Balance"
    )]
    erc1155_balance_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC20-Balance")]
    erc20_balance_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC20-Name")]
    erc20_name_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC20-Symbol")]
    erc20_symbol_header: HeaderName,

    #[structopt(
      env,
      long,
      value_name = "name",
      default_value = "X-Web3-ERC721-Balance"
    )]
    erc721_balance_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC721-Name")]
    erc721_name_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC721-Symbol")]
    erc721_symbol_header: HeaderName,

    #[structopt(
      env,
      long,
      value_name = "name",
      default_value = "X-Web3-ERC777-Balance"
    )]
    erc777_balance_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC777-Name")]
    erc777_name_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-ERC777-Symbol")]
    erc777_symbol_header: HeaderName,

    #[structopt(env, long, value_name = "name", default_value = "X-Web3-Signature")]
    signature_header: HeaderName,

    #[structopt(env, long, short = "k", value_name = "path")]
    secret_key_file: Option<PathBuf>,

    #[structopt(env, long, short = "K", value_name = "hex")]
    secret_key_data: Option<HexData>,

    #[structopt(env, long, short, value_name = "phrase", default_value = "totes-legit")]
    challenge: String,

    #[structopt(env, long, short = "u", value_name = "unit", default_value = "Wei")]
    balance_scale: BalanceScale,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str))]
    balance_minimum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str))]
    balance_maximum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc1155_balance_minimum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc1155_balance_maximum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc20_balance_minimum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc20_balance_maximum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc721_balance_minimum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc721_balance_maximum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc777_balance_minimum: Option<U256>,

    #[structopt(env, long, value_name = "amount", parse(try_from_str = U256::from_dec_str), )]
    erc777_balance_maximum: Option<U256>,

    #[structopt(env, long, value_name = "address")]
    erc1155_contract_address: Option<Address>,

    #[structopt(env, long, value_name = "address")]
    erc20_contract_address: Option<Address>,

    #[structopt(env, long, value_name = "address")]
    erc721_contract_address: Option<Address>,

    #[structopt(env, long, value_name = "address")]
    erc777_contract_address: Option<Address>,

    #[structopt(
      env,
      long,
      short = "S",
      takes_value = false,
      help = "provide signatures"
    )]
    provides_signatures: bool,

    #[structopt(
      env,
      long,
      short = "V",
      takes_value = false,
      help = "verify signatures and provide account addresses"
    )]
    provides_account_verification: bool,

    #[structopt(
      env,
      long,
      short = "B",
      takes_value = false,
      help = "provide account balances"
    )]
    provides_balances: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC1155 balances")]
    provides_erc1155_balance: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC20 balances")]
    provides_erc20_balance: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC20 names")]
    provides_erc20_name: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC20 symbols")]
    provides_erc20_symbol: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC721 balances")]
    provides_erc721_balance: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC721 names")]
    provides_erc721_name: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC721 symbols")]
    provides_erc721_symbol: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC777 balances")]
    provides_erc777_balance: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC777 names")]
    provides_erc777_name: bool,

    #[structopt(env, long, takes_value = false, help = "provide ERC777 symbols")]
    provides_erc777_symbol: bool,
  },
  Contract(contract::Command),
  Guide(guide::Command),
}

pub async fn run() -> WrappedResult<()> {
  match Options::from_args() {
    Options::Demo { listen } => {
      log::with_level(log::LevelFilter::Debug);
      crate::application::demo::server().listen(listen).await?;
    }
    Options::Units { .. } => {
      if let Some(max) = BalanceScale::VARIANTS.iter().map(|&s| s.len()).max() {
        for &variant in BalanceScale::VARIANTS {
          let scale = BalanceScale::from_str(&variant)?.scale();
          println!("{:<pad$} => {}", &variant, scale, pad = max);
        }
      };
    }
    Options::Contract(command) => command.execute().await?,
    Options::Guide(command) => command.execute()?,
    Options::Web3 {
      address_header,
      backend,
      balance_header,
      balance_maximum,
      balance_minimum,
      balance_scale,
      challenge: message,
      erc1155_balance_header,
      erc1155_balance_maximum,
      erc1155_balance_minimum,
      erc1155_contract_address,
      erc20_balance_header,
      erc20_balance_maximum,
      erc20_balance_minimum,
      erc20_contract_address,
      erc20_name_header,
      erc20_symbol_header,
      erc721_balance_header,
      erc721_balance_maximum,
      erc721_balance_minimum,
      erc721_contract_address,
      erc721_name_header,
      erc721_symbol_header,
      erc777_balance_header,
      erc777_balance_maximum,
      erc777_balance_minimum,
      erc777_contract_address,
      erc777_name_header,
      erc777_symbol_header,
      listen,
      provides_account_verification,
      provides_balances,
      provides_erc1155_balance,
      provides_erc20_balance,
      provides_erc20_name,
      provides_erc20_symbol,
      provides_erc721_balance,
      provides_erc721_name,
      provides_erc721_symbol,
      provides_erc777_balance,
      provides_erc777_name,
      provides_erc777_symbol,
      provides_signatures,
      secret_key_data,
      secret_key_file,
      signature_header,
      web3_rpc_url,
      ..
    } => {
      let secret_key = match (secret_key_data, secret_key_file) {
        (Some(data), _) => Some(SecretKey::from_slice(&data.0)?),
        (None, Some(path)) => Some(SecretKey::from_slice(&fs::read(path)?)?),
        (None, None) => None,
      };

      let balance_requirement = balance_requirement_from(balance_minimum, balance_maximum);

      let zero_address = Address::from_str("0x0000000000000000000000000000000000000000")?;

      let erc1155 = crate::application::proxy::ERC1155Config {
        balance_header: erc1155_balance_header,
        balance_requirement: balance_requirement_from(
          erc1155_balance_minimum,
          erc1155_balance_maximum,
        ),
        contract_address: match erc1155_contract_address {
          Some(address) => address,
          None => zero_address,
        },
        provides_balances: provides_erc1155_balance,
      };

      let erc20 = crate::application::proxy::ERC20Config {
        balance_header: erc20_balance_header,
        balance_requirement: balance_requirement_from(erc20_balance_minimum, erc20_balance_maximum),
        contract_address: match erc20_contract_address {
          Some(address) => address,
          None => zero_address,
        },
        name_header: erc20_name_header,
        provides_balances: provides_erc20_balance,
        provides_name: provides_erc20_name,
        provides_symbol: provides_erc20_symbol,
        symbol_header: erc20_symbol_header,
      };

      let erc721 = crate::application::proxy::ERC721Config {
        balance_header: erc721_balance_header,
        balance_requirement: balance_requirement_from(
          erc721_balance_minimum,
          erc721_balance_maximum,
        ),
        contract_address: match erc721_contract_address {
          Some(address) => address,
          None => zero_address,
        },
        name_header: erc721_name_header,
        provides_balances: provides_erc721_balance,
        provides_name: provides_erc721_name,
        provides_symbol: provides_erc721_symbol,
        symbol_header: erc721_symbol_header,
      };

      let erc777 = crate::application::proxy::ERC777Config {
        balance_header: erc777_balance_header,
        balance_requirement: balance_requirement_from(
          erc777_balance_minimum,
          erc777_balance_maximum,
        ),
        contract_address: match erc777_contract_address {
          Some(address) => address,
          None => zero_address,
        },
        name_header: erc777_name_header,
        provides_balances: provides_erc777_balance,
        provides_name: provides_erc777_name,
        provides_symbol: provides_erc777_symbol,
        symbol_header: erc777_symbol_header,
      };

      let config = crate::application::proxy::Config {
        address_header,
        backend,
        balance_header,
        balance_requirement,
        balance_scale: Some(balance_scale),
        challenge: message.as_bytes().to_vec(),
        erc1155,
        erc20,
        erc721,
        erc777,
        provides_account_verification,
        provides_balances,
        provides_signatures,
        secret_key,
        signature_header,
        web3_rpc_url,
      };

      log::with_level(log::LevelFilter::Debug);
      crate::application::proxy::server(config)
        .await?
        .listen(listen)
        .await?;
    }
  }

  Ok(())
}

fn balance_requirement_from(
  balance_minimum: Option<U256>,
  balance_maximum: Option<U256>,
) -> Option<BalanceRequirement> {
  match (balance_minimum, balance_maximum) {
    (Some(min), None) => Some(BalanceRequirement::AtLeast(min)),
    (None, Some(max)) => Some(BalanceRequirement::AtMost(max)),
    (Some(min), Some(max)) => Some(BalanceRequirement::Between(min, max)),
    (None, None) => None,
  }
}
